File: plugins/collector/event/configs/sshdconfig.go
package configs

import (
	"bufio"
	"collector/eventmanager"
	"collector/utils"
	"io"
	"os"
	"strings"
	"unicode"

	"github.com/chriskaliX/SDK"
	"github.com/chriskaliX/SDK/transport/protocol"
)

const (
	sshdConfig    = "/etc/ssh/sshd_config"
	SSHD_DATATYPE = 3002
)

var _ eventmanager.IEvent = (*Sshd)(nil)

type Sshd struct{}

func (Sshd) DataType() int {
	return SSHD_DATATYPE
}

func (Sshd) Name() string {
	return "sshdconfig"
}

func (n *Sshd) Flag() eventmanager.EventMode {
	return eventmanager.Periodic
}

func (Sshd) Immediately() bool { return false }

func (Sshd) Run(s SDK.ISandbox, sig chan struct{}) error {
	hash := utils.Hash()
	result := make(map[string]string, 7)
	var scan *bufio.Scanner
	file, err := os.Open(sshdConfig)
	if err != nil {
		return err
	}
	defer file.Close()
	// Default value of the configuration
	result["pubkey_authentication"] = "yes"
	result["passwd_authentication"] = "no"
	result["permit_emptypassword"] = "no"
	result["permit_rootlogin"] = "no"
	result["port"] = "22"
	result["max_authtries"] = "-1" // In my vm, it is 6

	scan = bufio.NewScanner(io.LimitReader(file, 1024*1024))
	for scan.Scan() {
		text := strings.TrimSpace(scan.Text())
		// skip
		if len(text) == 0 || text[:1] == "#" {
			continue
		}
		// Also, according to https://www.cyberciti.biz/faq/create-ssh-config-file-on-linux-unix/
		// "=" is also supported, which is ignored in Elkeid. Just a tidy problem, which can be used
		// in avoiding detection of ssh_config.
		// get PasswordAuthentication & PubkeyAuthentication Only
		fields := strings.FieldsFunc(text, func(c rune) bool {
			return unicode.IsSpace(c) || c == '='
		})
		if len(fields) == 2 {
			switch strings.TrimSpace(fields[0]) {
			case "PasswordAuthentication":
				result["passwd_authentication"] = strings.TrimSpace(fields[1])
			case "PubkeyAuthentication":
				result["pubkey_authentication"] = strings.TrimSpace(fields[1])
			case "PermitEmptyPasswords":
				result["permit_emptypassword"] = strings.TrimSpace(fields[1])
			case "MaxAuthTries":
				result["max_auth_tries"] = strings.TrimSpace(fields[1])
			case "PermitRootLogin":
				result["permit_rootlogin"] = strings.TrimSpace(fields[1])
			case "Port":
				result["Port"] = strings.TrimSpace(fields[1])
			}
		}
	}
	rec := &protocol.Record{
		DataType: SSHD_DATATYPE,
		Data: &protocol.Payload{
			Fields: result,
		},
	}
	rec.Data.Fields["package_seq"] = hash
	s.SendRecord(rec)
	return nil
}

func init() { addEvent(&Sshd{}) }

File: plugins/collector/src/event/configs/sshdconfig.rs
/// SshdConfig collector — data_type 3002.
/// Parses /etc/ssh/sshd_config for key security settings.
/// Mirrors Go's event/configs/sshdconfig.go.
use std::collections::HashMap;
use std::fs;
use std::io::{BufRead, BufReader, Read};

use anyhow::Result;
use sdk::Client;

use crate::event::{hash, make_record};

const DATA_TYPE: i32 = 3002;
const SSHD_CONFIG: &str = "/etc/ssh/sshd_config";

pub async fn run(client: &mut Client) -> Result<()> {
    let mut result: HashMap<String, String> = HashMap::new();
    // Defaults (mirrors Go's pre-filled values)
    result.insert("pubkey_authentication".into(), "yes".into());
    result.insert("passwd_authentication".into(), "no".into());
    result.insert("permit_emptypassword".into(),  "no".into());
    result.insert("permit_rootlogin".into(),      "no".into());
    result.insert("port".into(),                  "22".into());
    result.insert("max_authtries".into(),         "-1".into());

    if let Ok(f) = fs::File::open(SSHD_CONFIG) {
        let reader = BufReader::new(f.take(1024 * 1024));
        for line in reader.lines().flatten() {
            let text = line.trim().to_owned();
            if text.is_empty() || text.starts_with('#') { continue; }
            // Split by any whitespace or '=', skip empty tokens — mirrors Go's strings.FieldsFunc
            let fields: Vec<&str> = text
                .split(|c: char| c.is_whitespace() || c == '=')
                .filter(|s| !s.is_empty())
                .collect();
            if fields.len() != 2 { continue; }
            match fields[0] {
                "PasswordAuthentication" => { result.insert("passwd_authentication".into(), fields[1].to_owned()); }
                "PubkeyAuthentication"   => { result.insert("pubkey_authentication".into(), fields[1].to_owned()); }
                "PermitEmptyPasswords"   => { result.insert("permit_emptypassword".into(),  fields[1].to_owned()); }
                "MaxAuthTries"           => { result.insert("max_auth_tries".into(),         fields[1].to_owned()); }
                "PermitRootLogin"        => { result.insert("permit_rootlogin".into(),        fields[1].to_owned()); }
                "Port"                   => { result.insert("port".into(),                    fields[1].to_owned()); }
                _ => {}
            }
        }
    }

    result.insert("package_seq".into(), hash());
    let _ = client.send_record(&make_record(DATA_TYPE, result));
    Ok(())
}
